drivers/i3c/i3c_rtio.c

```diff
@@ -0,0 +1,254 @@
+ * Copyright (c) 2023 Intel Corporation
+ * Copyright (c) 2024 Meta Platforms
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/i3c.h>
+#include <zephyr/drivers/i3c/rtio.h>
+#include <zephyr/rtio/rtio.h>
+#include <zephyr/sys/mpsc_lockfree.h>
+#include <zephyr/sys/__assert.h>
+
+#define LOG_LEVEL CONFIG_I3C_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(i3c_rtio);
+
+const struct rtio_iodev_api i3c_iodev_api = {
+	.submit = i3c_iodev_submit,
+};
+
+struct rtio_sqe *i3c_rtio_copy(struct rtio *r, struct rtio_iodev *iodev, const struct i3c_msg *msgs,
+			       uint8_t num_msgs)
+{
+	__ASSERT(num_msgs > 0, "Expecting at least one message to copy");
+
+	struct rtio_sqe *sqe = NULL;
+
+	for (uint8_t i = 0; i < num_msgs; i++) {
+		sqe = rtio_sqe_acquire(r);
+
+		if (sqe == NULL) {
+			rtio_sqe_drop_all(r);
+			return NULL;
+		}
+
+		if (msgs[i].flags & I3C_MSG_READ) {
+			rtio_sqe_prep_read(sqe, iodev, RTIO_PRIO_NORM, msgs[i].buf, msgs[i].len,
+					   NULL);
+		} else {
+			rtio_sqe_prep_write(sqe, iodev, RTIO_PRIO_NORM, msgs[i].buf, msgs[i].len,
+					    NULL);
+		}
+		sqe->flags |= RTIO_SQE_TRANSACTION;
+		sqe->iodev_flags =
+			((msgs[i].flags & I3C_MSG_STOP) ? RTIO_IODEV_I3C_STOP : 0) |
+			((msgs[i].flags & I3C_MSG_RESTART) ? RTIO_IODEV_I3C_RESTART : 0) |
+			((msgs[i].flags & I3C_MSG_HDR) ? RTIO_IODEV_I3C_HDR : 0) |
+			((msgs[i].flags & I3C_MSG_NBCH) ? RTIO_IODEV_I3C_NBCH : 0) |
+			RTIO_IODEV_I3C_HDR_MODE_SET(msgs[i].hdr_mode) |
+			RTIO_IODEV_I3C_HDR_CMD_CODE_SET(msgs[i].hdr_cmd_code);
+	}
+
+	sqe->flags &= ~RTIO_SQE_TRANSACTION;
+
+	return sqe;
+}
+
+void i3c_rtio_init(struct i3c_rtio *ctx)
+{
+	k_sem_init(&ctx->lock, 1, 1);
+	mpsc_init(&ctx->io_q);
+	ctx->txn_curr = NULL;
+	ctx->txn_head = NULL;
+	ctx->iodev.api = &i3c_iodev_api;
+}
+
+/**
+ * @private
+ * @brief Setup the next transaction (could be a single op) if needed
+ *
+ * @retval true New transaction to start with the hardware is setup
+ * @retval false No new transaction to start
+ */
+static bool i3c_rtio_next(struct i3c_rtio *ctx, bool completion)
+{
+	k_spinlock_key_t key = k_spin_lock(&ctx->slock);
+
+	/* Already working on something, bail early */
+	if (!completion && ctx->txn_head != NULL) {
+		k_spin_unlock(&ctx->slock, key);
+		return false;
+	}
+
+	struct mpsc_node *next = mpsc_pop(&ctx->io_q);
+
+	/* Nothing left to do */
+	if (next == NULL) {
+		ctx->txn_head = NULL;
+		ctx->txn_curr = NULL;
+		k_spin_unlock(&ctx->slock, key);
+		return false;
+	}
+
+	ctx->txn_head = CONTAINER_OF(next, struct rtio_iodev_sqe, q);
+	ctx->txn_curr = ctx->txn_head;
+
+	k_spin_unlock(&ctx->slock, key);
+
+	return true;
+}
+
+bool i3c_rtio_complete(struct i3c_rtio *ctx, int status)
+{
+	/* On error bail */
+	if (status < 0) {
+		rtio_iodev_sqe_err(ctx->txn_head, status);
+		return i3c_rtio_next(ctx, true);
+	}
+
+	/* Try for next submission in the transaction */
+	ctx->txn_curr = rtio_txn_next(ctx->txn_curr);
+	if (ctx->txn_curr) {
+		return true;
```

Comment:
So in this case the sqe does not get completed neither with 'ok' nor with 'err'. Is this correct?

Reply:
I would assume so?? `i2c_rtio.c` is doing the exact same thing

Reply:
Yes yes, but still I don't get it. But the code seems well written, so there should be a reason I'm not 100% getting

Reply:
I'm not really sure either... I'm being rather a coding monkey :monkey_face: here and copied most of the `i2c_rtio.c`, but only understanding just enough to make it work with i3c. Perhaps @teburd would now as the original author?

Reply:
Transactions should result in one completion, either the transaction succeeds or fails. Akin to calling i2c_transfer you only know th result of the full operation not the individual sends/recvs